### PR TITLE
Add usePrevious react hook

### PR DESCRIPTION
**Pull request by sylvhama**
## Description

Fixes https://github.com/Shopify/web/issues/20093

`web` is already using this hook at several places. Instead of copy pasting it over and over I guess it would be nice to put it here.

More details about this hook: https://usehooks.com/usePrevious/

- [ ]  Patch: Bug/ Documentation fix (non-breaking change which fixes an issue or adds documentation)
- [x]  Minor: New feature (non-breaking change which adds functionality)
- [ ]  Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have added a changelog entry, prefixed by the type of change noted above

### CODE REVIEW

File: packages/react-hooks/README.md [hunk 1]
```diff
@@ -84,3 +84,14 @@ function MockComponent() {
   );
 }
 ```
+
+### `usePrevious()`
+
+This hook will storage the previous value of a given value.
```
Comment (loic-d) [inline packages/react-hooks/README.md]: will store

File: packages/react-hooks/src/hooks/previous.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import {useRef, useEffect} from 'react';
+
+export default function usePrevious<T>(value: T) {
+  const ref = useRef<T>();
+  useEffect(() => {
+    ref.current = value;
+  }, [value]);
+
+  return ref.current;
+}
```
Comment (loic-d) [inline packages/react-hooks/src/hooks/previous.ts]: Missing line break 💅

File: packages/react-hooks/README.md [hunk 3]
```diff
@@ -84,3 +84,14 @@ function MockComponent() {
   );
 }
 ```
+
+### `usePrevious()`
+
+This hook will store the previous value of a given value.
```
Comment (loic-d) [inline packages/react-hooks/README.md]: Should we say 'of a given variable'?

File: packages/react-hooks/src/hooks/tests/previous.test.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    return <div>{previousValue}</div>;
+  }
+
+  it('inits the previous value to undefined and displays it', () => {
+    const wrapper = mount(<MockComponent value={666} />);
+
+    expect(wrapper).toContainReactComponent('div', {children: undefined});
+  });
+
+  it('displays previous value after updating it', () => {
```
Comment (loic-d) [inline packages/react-hooks/src/hooks/tests/previous.test.tsx]: ... after updating value?

File: packages/react-hooks/README.md [hunk 5]
```diff
@@ -84,3 +84,14 @@ function MockComponent() {
...
 ```
+
+### `usePrevious()`
+
+This hook will store the previous value of a given variable.
+
+```tsx
+function MockComponent({value}) {
+  const previousValue = usePrevious(value);
+  return null;
```
Comment (cartogram) [inline packages/react-hooks/README.md]: I'd love to see a more realistic example here. Something that shows a very minimal real use-case.

File: packages/react-hooks/src/hooks/tests/previous.test.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import usePrevious from '../previous';
+
+describe('usePrevious', () => {
+  function MockComponent({value}) {
+    const previousValue = usePrevious(value);
+    return <div>{previousValue}</div>;
+  }
+
+  it('inits the previous value to undefined and displays it', () => {
```
Comment (cartogram) [inline packages/react-hooks/src/hooks/tests/previous.test.tsx]: maybe just `returns undefined initially`?

**Comment by sylvhama** [comment]
@cartogram @loic-d I have made some text updates based on your comments and I also created a better example. Let me know if you have further concerns :)

File: packages/react-hooks/README.md [hunk 7]
```diff
@@ -84,3 +84,19 @@ function MockComponent() {
...
+### `usePrevious()`
+
+This hook will store the previous value of a given variable.
+
+```tsx
+function Score({value}) {
+  const previousValue = usePrevious(value);
+  return (
+    <>
+      <p>{value}</p>
```
Comment (loic-d) [inline packages/react-hooks/README.md]: Nit but maybe we could have something like `<p>Current score: {value}</p>`

File: packages/react-hooks/README.md [hunk 8]
```diff
@@ -84,3 +84,19 @@ function MockComponent() {
...
+      {previousValue < value && <p>New record!</p>}
```
Comment (loic-d) [inline packages/react-hooks/README.md]: Nit but I would write the condition the other way around: "If the next scrore is greater than the previous one, we have a new record"

`value > previousValue && <p>We have a new record!</p>`

File: packages/react-hooks/src/hooks/tests/previous.test.tsx [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import {mount} from '@shopify/react-testing';
+
+import usePrevious from '../previous';
+
+describe('usePrevious', () => {
+  function MockComponent({value}) {
+    const previousValue = usePrevious(value);
```
Comment (loic-d) [inline packages/react-hooks/src/hooks/tests/previous.test.tsx]: Should we re-use your 'real-world' example in the test?

File: packages/react-hooks/README.md [hunk 10]
```diff
@@ -84,3 +84,19 @@ function MockComponent() {
...
+
+This hook will store the previous value of a given variable.
+
+```tsx
+function Score({value}) {
+  const previousValue = usePrevious(value);
+  return (
+    <>
+      <p>Current score: {value}</p>
+      {value > previousValue && <p>We have a new record!</p>}
```
Comment (ismail-syed) [inline packages/react-hooks/README.md]: Nit. This coding style feels a bit off from how we generally handle conditionally rendering content.  I would suggest pulling it out 

```tsx
const newRecord = value > previousValue ? <p>We have a new record!</p> : null;

return ( 
...
  {newRecord}
...
) 
```

Comment (sylvhama) [reply packages/react-hooks/README.md]: Done!

**Pull request merged**